tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,159 @@
+        my $cat_affinity_files_cmdline = "ssh root\@$guest \"cd /proc/irq && cat";
+        foreach my $irq_id (@nic_irqs_id) {
+            assert_script_run("ssh root\@$guest \"echo 1 > /proc/irq/$irq_id/smp_affinity\"");
+            $cat_affinity_files_cmdline .= " $irq_id/smp_affinity";
+        }
+        $cat_affinity_files_cmdline .= "\"";
+        # with irqbalance.service is enabled, SMP affinity will be auto balanced even if IRQs are bound to specified CPUs in SLE15SP4
+        # while in prior releases, such as SLE15SP3 or SLE12SP5, IRQs can be bound successfully until another irqbalance is run again
+        assert_script_run("ssh root\@$guest \"irqbalance --oneshot\"") if guest_is_sle($guest, '<=15-sp3');
+        my @affinity_after_binding = split('\n', script_output($cat_affinity_files_cmdline, 60));
+        foreach my $affinity (@affinity_after_binding) {
+            if ($affinity eq 1) {
+                # for SLE15SP3 and prior, IRQs are distributed on one cpu core ramdomly.
+                # So it is correct if some of them may happen to run on cpu0
+                # but it would not be correct if all IRQs were bound to cpu0 successfully.
+                if (guest_is_sle($guest, '<=15-sp3')) {
+                    record_soft_failure("One of the NIC IRQs was bound to cpu0, which may be not expected.");
+                }
+                # for SLE15SP4 (or newer release), IRQs are specified on all CPU cores
+                # fg. affinity value is '0xf' for machine with 4 CPU cores
+                else {
+                    die "It is not correct that one of the NIC IRQs was bound to cpu0 successfully with irqbalance enabled.";
+                }
+            }
+        }
+
+        # Generate xen guest network load
+        assert_script_run("ssh root\@$guest \"ping -f -c 500 download.opensuse.org\"");
+
+        # re-caculate the network interrupts distribution
+        # the increased NIC IRQs should be distributed in CPU cores in balance
+        my @total_irqs_on_cpu_after_network_download = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        my @increased_irqs_on_cpu;
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            $increased_irqs_on_cpu[$cpu_id] = $total_irqs_on_cpu_after_network_download[$cpu_id] - $initial_total_irqs_on_cpu[$cpu_id];
+            die("The value of network interrupts from download for CPU" . $cpu_id . " is 0") if $increased_irqs_on_cpu[$cpu_id] == 0;
+        }
+        record_info("IRQs of downloads distribution", "@increased_irqs_on_cpu");
+    }
+
+}
+
+#set up guest test environment to run irqbalance test
+sub prepare_guest_for_irqbalance {
+    my $vm_name = shift;
+
+    #4 or more vcpu is needed
+    my $nproc = script_output "virsh vcpucount $vm_name | grep current | grep config | awk '{ print \$3 }'";
+    if ($nproc < 4) {
```

Comment:
IMHO, if nproc > 1, it will be able to test irqbalance. Why do we need it to be 4? Besides, if there is strong reason for this, 'virsh set-vcpu xx 4 --live ' will be easier to satisfy this without worrying about changing the guest config(need restore after irq test)?

Reply:
After second thought on this, I would recommend to keep the guests with their original configured vcpu number if it is bigger than 1. The benefit is that we can cover different vcpu numbers and save the cost of changing guests and recovering them after test.

Reply:
irqbalance can be test with vcpu=2, but it is not a good test condition. because cpu0 always be the most-used one with/without irqbalance, it is hard to determine if the IRQs distributed in all CPUs. I think vcpu=4 or more are much better, but it can't be more then the IRQs generated by the NIC packets.

Regarding ''virsh set-vcpu xx 4 --live ' ', Leon has raised this issue as well. I have concerns about it even if I know it is easier:

1. ''virsh set-vcpu xx 4 --live ' only works with vm with <vcpu current=2>4</vcpu>. for VMs with other vpu configurations, the test will fail. while my test applies to any VMs.
2. cpu hotplug is another test. Involving another test to irqbalance test may introduce unnecessary troubles. the bug from vcpu test may block irqbalance test; besides, when we discribe the reproduce step in the case of an IRQ balance bug, ''virsh set-vcpu xx 4 --live' has to be listed, it is not a necessary step but it may introduce extra effort or explaination. I'd like a test simpler and purer.

Reply:
I see, reasonable too. Lucky that most guests are configured to have >=4 vcpus. 
BTW, since this guest vcpu confiig change is only done for guest with <4 vcpu. Then when multiple guests are tested with a mixure of > and < 4 vcpus, can restore_original_guests survive?

Reply:
Yes, the vm xml files are saved before irqbalance tests. Finally whatever the test fail or succeed, the VMs are re-defined with their original xml files.